crates/oxc_transformer/src/proposals/explicit_resource_management.rs

```diff
@@ -0,0 +1,768 @@
+        ForStatementLeft, ImportOrExportKind, ModuleExportName, Program, SimpleAssignmentTarget,
+        Statement, VariableDeclarationKind,
+    },
+};
+use oxc_ecmascript::BoundNames;
+use oxc_semantic::ScopeId;
+use oxc_semantic::{ScopeFlags, SymbolFlags};
+use oxc_span::{Atom, SPAN};
+use oxc_traverse::{BoundIdentifier, Traverse};
+use rustc_hash::FxHashMap;
+
+use crate::{Helper, context::TransformCtx};
+
+pub struct ExplicitResourceManagement<'a, 'ctx> {
+    ctx: &'ctx TransformCtx<'a>,
+
+    top_level_using: FxHashMap<Address, /* is await-using */ bool>,
+}
+
+impl<'a, 'ctx> ExplicitResourceManagement<'a, 'ctx> {
+    pub fn new(ctx: &'ctx TransformCtx<'a>) -> Self {
+        Self { ctx, top_level_using: FxHashMap::default() }
+    }
+}
+
+impl<'a> Traverse<'a> for ExplicitResourceManagement<'a, '_> {
+    /// Transforms ready for `enter_block_statement` to do the rest.
+    ///
+    /// ```ts
+    /// for await (using x of y) {}
+    /// ```
+    /// into
+    /// ```ts
+    /// for (const _x of y) {
+    ///     await using x = _x;
+    /// }
+    /// ```
+    fn enter_for_of_statement(
+        &mut self,
+        node: &mut oxc_ast::ast::ForOfStatement<'a>,
+        ctx: &mut oxc_traverse::TraverseCtx<'a>,
+    ) {
+        let for_of_stmt_scope_id = node.scope_id();
+        let ForStatementLeft::VariableDeclaration(decl) = &mut node.left else { return };
+        if !matches!(
+            decl.kind,
+            VariableDeclarationKind::Using | VariableDeclarationKind::AwaitUsing
+        ) {
+            return;
+        }
+        let variable_decl_kind = decl.kind;
+        decl.kind = VariableDeclarationKind::Const;
+
+        let variable_declarator = decl.declarations.first_mut().unwrap();
+        variable_declarator.kind = VariableDeclarationKind::Const;
+
+        let temp_id = ctx.generate_uid_based_on_node(
+            variable_declarator.id.get_binding_identifier().unwrap(),
+            for_of_stmt_scope_id,
+            SymbolFlags::ConstVariable | SymbolFlags::BlockScopedVariable,
+        );
+
+        let binding_pattern =
+            mem::replace(&mut variable_declarator.id, temp_id.create_binding_pattern(ctx));
+
+        let stmt = Statement::from(ctx.ast.declaration_variable(
+            SPAN,
+            variable_decl_kind,
+            ctx.ast.vec1(ctx.ast.variable_declarator(
+                SPAN,
+                variable_decl_kind,
+                binding_pattern,
+                Some(temp_id.create_read_expression(ctx)),
+                false,
+            )),
+            false,
+        ));
+
+        if let Statement::BlockStatement(body) = &mut node.body {
+            body.body.insert(0, stmt);
+        } else {
+            let old_body = ctx.ast.move_statement(&mut node.body);
+
+            let new_body = ctx.ast.vec_from_array([stmt, old_body]);
+            node.body = Statement::BlockStatement(ctx.ast.alloc_block_statement_with_scope_id(
```

Comment:
```suggestion
            node.body = ctx.ast.statement_block_statement_with_scope_id(
```

(but rebase on latest main first - Dunqing only added these shortcut APIs earlier this week)

Reply:
have a link to the code ref? i can't see this function

Reply:
nvm found it - `statement_block_with_scope_id`